devito/builtins.py

```diff
@@ -53,6 +99,165 @@ def smooth(f, g, axis=None):
         dv.Operator(dv.Eq(f, g.avg(dims=axis)), name='smoother')()
 
 
+def gaussian_smooth(f, sigma=1, _order=4, mode='reflect'):
+    """
+    Gaussian smooth function.
+    """
+    class ObjectiveDomain(dv.SubDomain):
+
+        name = 'objective_domain'
+
+        def __init__(self, lw):
+            super(ObjectiveDomain, self).__init__()
+            self.lw = lw
+
+        def define(self, dimensions):
+            return {d: ('middle', self.lw, self.lw) for d in dimensions}
+
+    def fset(f, g):
+        indices = [slice(lw, -lw, 1) for _ in g.grid.dimensions]
+        slices = (slice(None, None, 1), )*len(g.grid.dimensions)
+        if isinstance(f, np.ndarray):
+            f[slices] = g.data[tuple(indices)]
+        elif isinstance(f, dv.Function):
+            f.data[slices] = g.data[tuple(indices)]
+        else:
+            raise NotImplementedError
+
+    lw = int(_order*sigma + 0.5)
+
+    # Create the padded grid:
+    objective_domain = ObjectiveDomain(lw)
+    try:
+        shape_padded = np.array(f.grid.shape) + 2*lw
+    except AttributeError:
+        shape_padded = np.array(f.shape) + 2*lw
+    grid = dv.Grid(shape=shape_padded, subdomains=objective_domain)
+
+    f_c = dv.Function(name='f_c', grid=grid, space_order=2*lw,
+                      coefficients='symbolic', dtype=np.int32)
+    f_o = dv.Function(name='f_o', grid=grid, coefficients='symbolic', dtype=np.int32)
+
+    weights = np.exp(-0.5/sigma**2*(np.linspace(-lw, lw, 2*lw+1))**2)
+    weights = weights/weights.sum()
+
+    additional_expressions = {}
```

Comment:
nitpicking: for this sort of things, throughout devito we use the conventional name "mapper"